components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+                UPDATE_SUBSCRIBER_MODULE_SQL);
+             NamedPreparedStatement updateSubscriberPropertiesPrepStmt = new NamedPreparedStatement(connection,
+                     UPDATE_SUBSCRIBER_PROPERTIES_SQL)) {
+
+            if (!oldHolder.getModuleName().equalsIgnoreCase(holder.getModuleName())) {
+                updateSubscriberPrepStmt.setString(ENTITLEMENT_MODULE_NAME, holder.getModuleName());
+                updateSubscriberPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+                updateSubscriberPrepStmt.setInt(TENANT_ID, tenantId);
+                updateSubscriberPrepStmt.executeUpdate();
+            }
+
+            // Update the property values of an existing subscriber
+            for (PublisherPropertyDTO dto : propertyDTOs) {
+
+                if (StringUtils.isNotBlank(dto.getId()) && StringUtils.isNotBlank(dto.getValue())) {
+                    PublisherPropertyDTO propertyDTO;
+                    propertyDTO = oldHolder.getPropertyDTO(dto.getId());
+                    if (propertyDTO != null && !propertyDTO.getValue().equalsIgnoreCase(dto.getValue())) {
+                        updateSubscriberPropertiesPrepStmt.setString(PROPERTY_VALUE, dto.getValue());
+                        updateSubscriberPropertiesPrepStmt.setString(PROPERTY_ID, dto.getId());
+                        updateSubscriberPropertiesPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+                        updateSubscriberPropertiesPrepStmt.setInt(TENANT_ID, tenantId);
+                        updateSubscriberPropertiesPrepStmt.addBatch();
+                    }
+                }
+            }
+            updateSubscriberPropertiesPrepStmt.executeBatch();
+            IdentityDatabaseUtil.commitTransaction(connection);
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new EntitlementException("Error while updating subscriber details", e);
+        } finally {
+            IdentityDatabaseUtil.closeConnection(connection);
+        }
+    }
+
+    /**
+     * Sets the base64 encoded secret value of the secret subscriber properties, if it has been updated.
+     *
+     * @param holder    publisher data holder
+     * @param oldHolder existing publisher data holder
+     */
+    private void populateProperties(PublisherDataHolder holder, PublisherDataHolder oldHolder) {
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+
+        for (PublisherPropertyDTO dto : propertyDTOs) {
+            if (StringUtils.isNotBlank(dto.getId()) && StringUtils.isNotBlank(dto.getValue()) && dto.isSecret()) {
+                PublisherPropertyDTO propertyDTO = null;
+                if (oldHolder != null) {
+                    propertyDTO = oldHolder.getPropertyDTO(dto.getId());
+                }
+                if (propertyDTO == null || !propertyDTO.getValue().equalsIgnoreCase(dto.getValue())) {
```

Comment:
The `oldHolder` passed to this level have its secrets in encrypted manner. Ref [1].

And `propertyDTO` values are in plain text manner and those are getting encrypted inside this if block.
So, this compare encrypted value with a plaintext value, which leads to false regardless of its content.

Fixing as mentioned in [2] will solve this issue. But it will cause for redundant updates at [3].

[1] https://github.com/wso2/carbon-identity-framework/pull/5826/files#diff-50b8fca7333ced1dc9278978245c1128f4d26e7ed4c0449bd944ea1bf6b25608R163

[2] https://github.com/wso2/carbon-identity-framework/pull/5826/files#r1699815934

[3] https://github.com/wso2/carbon-identity-framework/pull/5826/files#diff-50b8fca7333ced1dc9278978245c1128f4d26e7ed4c0449bd944ea1bf6b25608R293

Reply:
The propryDTO value in the new holder will still have the encrypted value if it was not updated. So redundant updates won't happen